Support DVS input from files

hunse opened the pull request:
This allows loading DVS input from either `.aedat` or `.events` files and sending it to the chip.

hunse commented on docs/examples/dvs-from-file.ipynb:
```diff
@@ -0,0 +1,318 @@
+    "\n",
+    "height = 180 // pool[0]\n",
+    "width = 240 // pool[1]\n",
+    "gain = 101\n",
```
I just had it at 100 at first, and then found that 101 gives slightly better results. I think it's because 100 can often result in values that are right around the firing threshold, where as 101 more consistently pushes things over the firing threshold (e.g. if a neuron has a firing threshold of 200, an input of 100 might put it at but not over the firing threshold after two timesteps, causing it to fire every three timesteps, whereas with the higher gain, the input would be 101 and thus consistently put it over).

hunse commented on docs/examples/dvs-from-file.ipynb:
```diff
@@ -0,0 +1,318 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pool = (10, 10)\n",
```
Just to keep everything fairly small. I think I wrote this example before we had block splitting, so I wanted to only have to make one ensemble to keep it straightforward. Now with block splitting, it shouldn't be too hard if we wanted to make it bigger.

drasmuss replied:
I think it's fine either way. I guess the only question would be whether we want to add some brief prose telling people why they'd want to use this (like "if you find yourself running out of axons, you might want to use pooling" or something like that).

hunse commented on docs/examples/dvs-from-file.ipynb:
```diff
@@ -0,0 +1,318 @@
+    "\n",
+    "    for k, e in enumerate(ensembles):\n",
+    "        u_channel = u[k :: u.polarity]\n",
+    "        nengo.Connection(u_channel, e.neurons, transform=1.0 / np.prod(pool))\n",
```
We could. I would want things to be consistent with the live DVS (when we add that), and there it might be a bit harder to do that. The way that live DVS works is you basically have one neuron per input pixel that gets the input from the DVS and relays the spikes; you can then arbitrarily control the axons coming out of those neurons to direct the DVS input. So for pooling, I just have multiple axons go to the same target synapses. To do the scaling, there might be a bit of work necessary in the builder to find which synapses things are being redirected to, and scale them accordingly. Probably not terribly difficult, but certainly a bit more work.

The scaling would certainly be more like average/max pooling if we had the scaling factor in there, and people who are coming from an ML background would be more familiar with those.

drasmuss replied:
I don't think it's that critical, so if it makes the implementation simpler to leave it out I'd go with that. Maybe just adding a comment in the notebook explaining why scaling is being applied, to make it stand out a bit more for anyone reading the example (so they'll think about whether they want to be adding that in their own code).

hunse commented on docs/examples/dvs-from-file.ipynb:
```diff
@@ -0,0 +1,318 @@
+    "\n",
+    "plt.tight_layout()\n",
+    "plt.show()"
+   ]
```
I think I was worried about if the animation didn't work for some people (you do need ffmpeg), but it is working in the online docs, so I guess they can always refer to that if need be.

drasmuss replied:
I didn't have ffmpeg installed originally and I just got a hard error somewhere (I don't remember exactly), so I don't think it really fails gracefully regardless.

hunse commented on nengo_loihi/__init__.py:
```diff
@@ -10,6 +10,7 @@
 from . import builder
 
 # Import into top-level namespace
+from .builder.inputs import DVSFileChipNode
```
My reasoning for putting it in the main namespace is that a) the main namespace doesn't seem over-full, b) it already has DVS in the name, so `nengo_loihi.dvs.DVSFileChipNode` seems a bit redundant, and c) it's not being "hidden" among the other things in the `nengo_loihi.dvs` namespace that most users will never need to touch.

drasmuss replied:
I guess it depends how many users we expect to be using this. I was thinking that it'd be pretty uncommon, so having it a bit hidden was more of an advantage than a disadvantage. But maybe there's a decent overlap between people using Loihi and people who have DVS cameras 🤷.

hunse commented on nengo_loihi/builder/inputs.py:
```diff
@@ -85,3 +91,116 @@ class PESModulatoryTarget:
+        self.rel_time = rel_time
+
+        self.dvs_height = 180
+        self.dvs_width = 240
```
So these are standard for the DVS/Davis sensor. But one user was asking about different sizes. We might as well make them configurable; we just want to make it clear that you should only be changing it if you actually have a sensor/file format that is a different size (i.e., not just because you want e.g. a smaller image for your model). I'm also not sure if NxSDK supports live DVS in other sizes (my impression from before was that they didn't, but that could have changed or I could just be wrong), but I'm OK with there being some features on the file DVS that aren't supported on live DVS.

hunse commented on nengo_loihi/builder/inputs.py:
```diff
@@ -47,10 +52,11 @@ class ChipReceiveNode(Node):
         self.spike_target = None
-        super().__init__(self.update, size_in=0, size_out=size_out, label=label)
+        output = self.update if output is None else output
+        super().__init__(output, size_in=0, size_out=size_out, label=label)
```
So I started with just overwriting `.update`, but there was some benefit to having it as a process that I honestly can't remember right now.

I would be OK with making this not be a subclass of `ChipReceiveNeurons`. I agree that it doesn't completely fit with the original purpose of that class. It can't just be a normal `Node`, because the splitter needs to know to put it on the chip, which is part of the reason I subclassed it in the first place. I would be okay having a general subclass of `ChipNode`, which `DVSFileChipNode` is a subclass of, and `ChipReceiveNode` and `ChipReceiveNeurons` are also subclasses of. Does that sound OK?

drasmuss replied:
Rather than adding `ChipNode`, I might just have two conditions in the splitter check instead (like `if isinstance(node, ChipReceiveNode) or isinstance(node.output, DVSFileImageProcess)`). It makes it a bit more awkward in the splitter, but that seems like less added complexity overall than adding a new class to the hierarchy.

drasmuss replied:
Could we get rid of `DVSFileChipNode` entirely in that case, and just have `DVSFileImageProcess` (combined with regular Nodes)?

hunse replied:
My idea behind the new class is that we might have more such nodes in the future. But maybe better to cross that bridge when we come to it.

The idea of doing it all as a process, and just checking for that as the output, and then getting rid of `DVSFileChipNode` entirely, is an interesting one. One potential downside is that it could cause a different way of doing things with files vs live DVS, since I'm not sure if we'd want to put in the work to allow the live DVS node to work as a process (or even in core Nengo at all), though it would certainly be cool if we did. 

It would also make things a bit more complicated in the builder, since right now I just have a different builder for `DVSFileChipNode` versus `ChipReceiveNode`. We currently don't have any special process builders in NengoLoihi. I guess we could do it in a general `Node` builder, and we'd know that any Nodes that get through to that point are being built on-chip, since otherwise they'd use the core Nengo build system. So it's definitely doable, and maybe not even too bad, but certainly a different way of going about things. Conceptually, I think I like having the separate Node type a little bit more, because it makes it clearer to users where the action is happening (that was my idea behind the name `DVSFileChipNode`, is that it indicates that the Node is happening on the chip, unlike normal Nodes). So if we move to the process, then we lose that. That would be my main argument for sticking with the current system.

drasmuss replied:
If it wouldn't be too much extra work to have the live DVS work as a process in NengoCore that'd definitely be cool (and generally useful across the whole ecosystem).

Another possibility is that we could have a `DVSFileChipNode` helper function (which under the hood would just be doing `Node(DVSFileImageProcess)`). That'd give the more visible distinction in the code/examples that this is a special kind of Node we're making (which I agree is useful), but allow us to keep things simpler under the hood.

hunse replied:
I'm not sure how much extra work it would be to have a live DVS process. I think you'd need to be running something like jAER in the background to actually interface with the DVS, and then connect to jAER in Python. I've always just used jAER to save files that I load, I've never done it live. @tcstewar might have experience with this. Anyway, it's a problem for future us.

A nice thing about having `ChipNode` is that it fits well into our pluggable builder system. So if somebody had some other spike data they want to pipe in, they can make their own `ChipNode` subclass that reads their file format and add a builder that puts the spikes in a `SpikeInput`. With processes, this wouldn't be possible, since they'd also have to add to the splitter that this type of process should be built by NengoLoihi, not Nengo core. To me, it seems like a clear way to group all special nodes that we want built by NengoLoihi.

That said, if you're really tied to the process idea, I could go for it. I like the idea of the `DVSFileChipNode` helper function in that case. It is true that the stuff in the process (as well as all the general DVS file code supporting it) is not NengoLoihi specific, so it's something that we could one day move to Nengo or NengoExtras, and the process model would fit much better with that.

So my preference would be to add the `ChipNode`, keep `DVSFileChipNode` as a subclass and hide the process as an implementation detail. Then down the road, if we add/move some of this functionality to Nengo/NengoExtras, we can do it as a process there. But as I said, I don't think it's a big deal either way.

tbekolay replied:
Part of the intention of the `Process` base class is to make things work with the pluggable builder system; I never intended for `Node` to be subclassed. I would rather it not work but I know that we programmed defensively enough that it works and the consensus didn't want to limit things for reasons like this, but I think there should not be reasons to subclass `Node`, it should be done with `Process` instead. If we had more resources for NengoLoihi, I would make `ChipNode` and other node subclasses into processes and refactor the splitter accordingly.

hunse replied:
So my question is what's an extensible way for users to indicate that a node should be built by the NengoLoihi build system rather than the core Nengo builder; i.e., how does a user tell the splitter that their process type should be put on the Loihi? If we can find a way to do that, then I agree that the process system is better.

Maybe we need a process subtype that the splitter automatically puts on chip? i.e. a `ChipProcess` rather than a `ChipNode`?

Or we could do something with the config system, though I feel like it's not ideal for this particular case (though maybe I'm just not able to imagine a good solution with it).

hunse commented on nengo_loihi/builder/inputs.py:
```diff
@@ -85,3 +91,116 @@ class PESModulatoryTarget:
+        self.size = self.height * self.width * self.polarity
+
+        # for outputting images if the node is run in Nengo
+        output = DVSFileImageProcess(self)
```
So my first inclination was that this makes a lot of sense, but when I started doing it, it made less sense.

The `DVSFileImageProcess` isn't something that we want users to be instantiating themselves. It's basically just a way to do DVS input as frames in Nengo core, and do it with the right `dt` (as opposed to modifying `.update`, which doesn't know `dt`). We always want users to use `DVSFileChipNode`, since that works properly in both Nengo core and NengoLoihi (if they did it with `DVSFileImageProcess`, and then started using the NengoLoihi simulator, it would make frames on the host, rather than just sending spikes directly to the chip, which is almost certainly not what they want).

I _guess_ it could be possible that a user would want a process to turn a DVS file into frames that they can run independently as a process, rather than having it as the output to a Node. That's the one case where I can think it would be useful for a user to make a `DVSFileImageProcess`.

I think we should make `DVSFileImageProcess` either hidden or a member of `DVSFileChipNode` (or both), to indicate that it's not meant to be instantiated by users. I would do that whether we move all the DVS parameters to `DVSFileImageProcess` or not. And I think that if we do that, then it's not really a problem to have the reference back to the node, so I wouldn't bother changing things. (The change means that you end up having all the DVS arguments on `DVSFileChipNode.__init__` and then passing them straight through to `DVSFileImageProcess.__init__`.)

drasmuss replied:
If we made the other change I suggested above (checking for the Process in the splitter, rather than the node type), then we wouldn't have that issue right? Because it'd be fine if users didn't use `DVSFileChipNode`, just a regular `Node(DVSFileImageProcess)` would work in NengoCore or NengoLoihi.

hunse commented:
Ok, I think I've addressed all of the comments. I ended up making a `ChipProcess` class; any nodes with a process subclassed from this will be put on the chip. This keeps things cleaner in the builder (we care about this in the connection builder, in addition to in the splitter). It also makes it possible for users to add their own chip processes, and makes it easier for us in the future if we want to add more.

I got rid of `DVSFileChipNode` entirely, and renamed the process to `DVSFileChipProcess` so it should be clear that it's sending spikes directly to the chip (I also made a note about this in the example).

drasmuss reviewed (approved):
Fixups all look good to me!
